File: static_site/src/types/etf.ts
export interface EtfInfo {
  code: string;
  description: string;
  distributions: EtfDistribution[];
  divided_info?: EtfDivideInfo[];
}

export interface EtfDivideInfo {
  date: string; // 分割/反分割生效日
  ratio: number; // 分割比率：比如 0.2 代表 5 股合 1 ， 10 代表 1 股切割為 10 股
}

export interface EtfDistribution {
  per_share: number;
  declared_date: string;
  ex_date: string;
  payable_date: string;
  roc: number;
}

export interface TwseEtfInfo {
  date?: string;
  code: string;
  name: string;
  company?: string;
  target?: string;
  market?: string;
  ex_date?: string;
  share?: number;
  total_ex_count: number;
  success_fill_count: number;
  win_rate: number;
  avg_fill_days: number;
  range_position: number;
  latest_close: number;
  ma5: number;
  ma20: number;
  ma20_bias_rate: number;
}

export interface TwseEtfShare extends TwseEtfInfo {
  // 用於即將除息
  code: string;
  ex_date: string;
  name: string;
  share: number;
  pre_ex_close_price: number;
  yield_rate: number;
  filled_date: string;
  filled_close_price: number;
  filled_days: number;
  filled_trade_days: number;
  payable_date: string;
  distribution: number;
}

export interface TwseEtfShareDetail {
  // 用於主列表
  stats: TwseEtfShare[];
  win_rate: {
    total_ex_count: number;
    success_fill_count: number;
    win_rate: number;
    avg_fill_days: number;
  };
}

export interface TwseETFTicker {
  date: string;
  open: number;
  max: number;
  min: number;
  close: number;
  range_position_20: number;
  range_position_60: number;
  range_position_120: number;
  ma5: number;
  ma20: number;
  ma60: number;
  ma120: number;
  volume: number;
  trading_money: number;
  trading_turnover: number;
}

File: model/entity/opendata/etf/etf.go
package etf

import (
	"time"

	"faryne.dev/model/enum"
)

type ETF struct {
	Date             string           `json:"date,omitempty" gorm:"column:publish_date"` // 發行日，抓回來時要把「.」轉換為「-」
	Code             string           `json:"code" gorm:"column:code"`
	Name             string           `json:"name"`
	Company          string           `json:"company,omitempty"`
	Target           string           `json:"target,omitempty"`
	Market           enum.StockMarket `json:"market,omitempty"` // twse / otc
	CreatedAt        time.Time        `json:"-" gorm:"column:created_at"`
	UpdatedAt        time.Time        `json:"-" gorm:"column:updated_at"`
	TotalExCount     int64            `json:"total_ex_count"`                          // 總共除權除息次數
	SuccessFillCount int64            `json:"success_fill_count"`                      // 成功填息次數
	WinRate          float64          `json:"win_rate"`                                // 勝率，  0 <= N <= 100
	AvgFillDays      float64          `json:"avg_fill_days"`                           // 平均填息日
	RangePosition    float64          `json:"range_position"`                          // 近一個月收盤價區間位置
	LatestClose      float64          `json:"latest_close" gorm:"column:latest_close"` // 最新收盤價
	MA5              float64          `json:"ma5" gorm:"column:ma5"`
	MA20             float64          `json:"ma20" gorm:"column:ma20"`
	MA20BiasRate     float64          `json:"ma20_bias_rate" gorm:"column:ma20_bias_rate"` // 收盤價相對 MA20 乖離率
	ExDate           time.Time        `json:"ex_date" gorm:"->"`
	Share            float64          `json:"share" gorm:"->"`
	PayableDate      string           `json:"payable_date" gorm:"->"`
}

func (e *ETF) TableName() string {
	return "etf_codes"
}

type Ticker struct {
	Date             string    `json:"date" gorm:"column:ticker_date"`
	Code             string    `json:"code" gorm:"column:code"`
	Open             float64   `json:"open"`
	Max              float64   `json:"max"`
	Min              float64   `json:"min"`
	Close            float64   `json:"close"`
	RangePosition20  float64   `json:"range_position_20" gorm:"column:range_position_20"`
	RangePosition60  float64   `json:"range_position_60" gorm:"column:range_position_60"`
	RangePosition120 float64   `json:"range_position_120" gorm:"column:range_position_120"`
	MA5              float64   `json:"ma5" gorm:"column:ma5"`
	MA20             float64   `json:"ma20" gorm:"column:ma20"`
	MA60             float64   `json:"ma60" gorm:"column:ma60"`
	MA120            float64   `json:"ma120" gorm:"column:ma120"`
	Volume           int64     `json:"volume"`
	TradingMoney     int64     `json:"trading_money" gorm:"column:trading_money"`
	TradingTurnover  int64     `json:"trading_turnover" gorm:"column:trading_turnover"`
	CreatedAt        time.Time `json:"-" gorm:"column:created_at"`
	UpdatedAt        time.Time `json:"-" gorm:"column:updated_at"`
}

func (t *Ticker) TableName() string {
	return "etf_tickers"
}

type Share struct {
	Id                int64     `json:"id"`
	Code              string    `json:"code" gorm:"column:code"`
	ExDate            string    `json:"ex_date"`
	PayableDate       string    `json:"payable_date"`
	Share             float64   `json:"share"`
	CreatedAt         time.Time `json:"-" gorm:"column:created_at"`
	UpdatedAt         time.Time `json:"-" gorm:"column:updated_at"`
	ExTickerPrice     float64   `json:"ex_ticker_price" gorm:"column:ex_ticker_price"`         // 除權息前收盤價，從 etf_tickers 找
	YieldRate         float64   `json:"yield_rate"`                                            // 單次殖利率
	FilledDate        string    `json:"filled_date"`                                           // 填息日
	FilledTickerPrice float64   `json:"filled_ticker_price" gorm:"column:filled_ticker_price"` // 填息日收盤價，從 etf_ticker 找
	FilledDays        int64     `json:"filled_days"`                                           // 填息天數（日曆天數）
	FilledTradeDays   int64     `json:"filled_trade_days"`                                     // 填息交易日數（交易日）
}

type ShareWithETFAndStats struct {
	ETF   `json:"etf"`
	Stats []Share `json:"stats"`
}

func (s *Share) TableName() string {
	return "etf_shares"
}

type MonthlyPrice struct {
	Id        int64     `json:"id" gorm:"column:id;primaryKey;autoIncrement"`
	Code      string    `json:"code" gorm:"column:code" validate:"required"`
	Year      int       `json:"year" gorm:"column:year" validate:"required"`
	Month     int       `json:"month" gorm:"column:month" validate:"required"`
	AvgPrice  float64   `json:"avg_price" gorm:"column:avg_price" validate:"required"`
	CreatedAt time.Time `json:"created_at" gorm:"column:created_at"`
	UpdatedAt time.Time `json:"updated_at" gorm:"column:updated_at"`
}

func (m *MonthlyPrice) TableName() string {
	return "etf_monthly_prices"
}
